// nightwatch-allure/lib/StepWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class StepWrapper {
    constructor(reporter, step) {
        this.reporter = reporter;
        this.step = step;
    }
    startStep(name) {
        const step = this.step.startStep(name);
        this.reporter.pushStep(step);
        return new StepWrapper(this.reporter, step);
    }
    setStatus(status) {
        this.step.status = status;
    }
    setDescription(description) {
        this.step.description = description;
    }
    setDetailsTrace(detailsTrace) {
        this.step.detailsTrace = detailsTrace;
        this.step.statusDetails = { message: "Error", trace: detailsTrace };
    }
    setDetailsMessage(detailsMessage) {
        this.step.detailsMessage = detailsMessage;
    }
    addAttachment(name, content, type) {
        const file = this.reporter.writeAttachment(content, type);
        this.step.addAttachment(name, type, file);
    }
    endStep(timeMs = 0) {
        this.reporter.popStep();
        this.step.endStep(timeMs);
    }
    run(body) {
        return this.step.wrap(body)();
    }
}
exports.StepWrapper = StepWrapper;
